lib/libzutil/zutil_import.c

```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
 
 	if (lio_listio(LIO_WAIT, aiocbps, VDEV_LABELS, NULL) != 0) {
 		int saved_errno = errno;
-
-		if (errno == EAGAIN || errno == EINTR || errno == EIO) {
-			/*
-			 * A portion of the requests may have been submitted.
-			 * Clean them up.
-			 */
-			for (l = 0; l < VDEV_LABELS; l++) {
-				errno = 0;
-				int r = aio_error(&aiocbs[l]);
-				if (r != EINVAL)
+		errno = 0;
+		int do_slow = 0;
+		error = -1;
+
+		for (l = 0; l < VDEV_LABELS; l++) {
+			switch (aio_error(&aiocbs[l])) {
+				case 0:
+					break;
+				case -1:
+					break;
+				case EOPNOTSUPP:
+					do_slow = 1;
+				case EIO:
```

Comment:
The EIO, EINTR, and EAGAIN error codes might be returned by `lio_listio`, but they should not be returned by `aio_error`.

Reply:
The documentation for aio_error on Linux says it can spit out ~anything that read could, if the operation failed, unless I'm misreading it?

Reply:
An EIO from aio_read means something very different from an EIO from lio_listio.  Only in the latter case does it mean that "A portion of the requests may have been submitted".  If aio_read returns EIO, it means that the disk actually wasn't readable.  Also, you should probably handle `ENOSYS` just like `EOPNOTSUPP`.  That won't happen on FreeBSD, but might on other platforms.

Reply:
Sure; I think I've got a reasonable workflow for this done; let me give it a whirl through some testing before I push it...

Reply:
This reply is kind of redundant with my one above, but I have refactored how this code flow works in response to this; if you have a chance, PTAL and let me know if I've missed something again.